### ISSUE TITLE
New NanoKVM - /dev/mmcblk0p3 - no filesytem

### ISSUE THREAD

**Issue by madworm** [sipeed/NanoKVM#76]
Just letting you know that I've received a "fresh" NanoKVM today

* /dev/mmcblk0p3 had no filesystem at all - according to the wiki it should be exFAT (I could fix that with mkfs.exfat, now works with usb)
* /etc/fstab doesn't have a mount point for it pointing to /data (but it gets mounted anyway after creating the exfat fs)

**Comment by polyzium** [comment]
This must be the data partition that is automatically created when the KVM boots. This happens whenever the firmware is reflashed, so I assume these little guys are mass-produced and were probably never turned on at the factory.

See https://github.com/sipeed/NanoKVM/issues/7#issuecomment-2288903849 for reflashing the firmware.

**Comment by madworm** [comment]
Well, the partition was there, but no exFAT had been created. I dumped the partition with dd and it was all zeros.

After creating the filesystem myself it worked.

Version 2.0.9

**Comment by neoKushan** [comment]
Yeah I got a new one a couple of days ago and had to re-image the included MicroSD card, as it wasn't working correctly.

**Comment by madworm** [comment]
What method did you use for that?

I tried step #5 from this, but nothing happened. 

`https://wiki.sipeed.com/hardware/en/kvm/NanoKVM/user_guide.html`

I copied the decompressed image file to the root of /dev/mmcblk0p2

The wiki makes it sound like upon reboot something is supposed to happen with it, but it just stays there. Or do I need to write the image to the sd-card with dd ?

**Comment by vodkapmp** [comment]
> The wiki makes it sound like upon reboot something is supposed to happen with it, but it just stays there. Or do I need to write the image to the sd-card with dd ?

Yeah, use dd to write it to the SD and you should be good

**Comment by Z2Z-GuGu** [comment]
@madworm Hi, we have implemented partition formatting and data mounting in/etc/init.d/S01fs. Has this been helpful for your question?

**Comment by NightowlKr** [comment]
"I hope my issue report #242 is helpful."

**Comment by mr-manuel** [comment]
@Z2Z-GuGu this is still a thing with v1.4.0 and a new device. The partition is not mounted. It seems that it needs a second reboot to do so. I will make a few more tests in the next days and report back.

See also https://github.com/sipeed/NanoKVM/issues/63 and https://github.com/sipeed/NanoKVM/issues/130

```bash
# cat /proc/partitions
major minor  #blocks  name

 179        0   31260672 mmcblk0
 179        1      16384 mmcblk0p1
 179        2    7983616 mmcblk0p2
 179        3   23260160 mmcblk0p3

# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/mmcblk0p2            7.5G      3.0G      4.1G  42% /
devtmpfs                 77.7M         0     77.7M   0% /dev
tmpfs                    79.0M         0     79.0M   0% /dev/shm
tmpfs                    79.0M     30.4M     48.6M  39% /tmp
tmpfs                    79.0M    128.0K     78.9M   0% /run
/dev/mmcblk0p1           16.0M     11.5M      4.5M  72% /boot

# mount
/dev/mmcblk0p2 on / type ext4 (rw,relatime)
proc on /proc type proc (rw,relatime)
sysfs on /sys type sysfs (rw,relatime)
devtmpfs on /dev type devtmpfs (rw,relatime,size=79572k,nr_inodes=19893,mode=755)
proc on /proc type proc (rw,relatime)
devpts on /dev/pts type devpts (rw,relatime,gid=5,mode=620,ptmxmode=666)
tmpfs on /dev/shm type tmpfs (rw,relatime,mode=777)
tmpfs on /tmp type tmpfs (rw,relatime)
tmpfs on /run type tmpfs (rw,nosuid,nodev,relatime,mode=755)
/dev/mmcblk0p1 on /boot type vfat (rw,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro)
configfs on /sys/kernel/config type configfs (rw,relatime)
debugfs on /sys/kernel/debug type debugfs (rw,relatime)

# cat /etc/fstab
# <file system> <mount pt>      <type>  <options>       <dump>  <pass>
/dev/root       /               ext2    rw,noauto       0       1
proc            /proc           proc    defaults        0       0
devpts          /dev/pts        devpts  defaults,gid=5,mode=620,ptmxmode=0666   0       0
tmpfs           /dev/shm        tmpfs   mode=0777       0       0
tmpfs           /tmp            tmpfs   mode=1777       0       0
tmpfs           /run            tmpfs   mode=0755,nosuid,nodev  0       0
sysfs           /sys            sysfs   defaults        0       0
```

**Comment by mr-manuel** [comment]
Very odd. I flashed 9 new SD cards, and the problem appeared only on the very first one...

**Comment by dormancygrace** [comment]
Hi! I've included a fix for this in [NanoKVM OS](https://github.com/dormancygrace/NanoKVM-OS), my independent community firmware for SG2002-based NanoKVM Cube and PCIe.

Data provisioning now tracks an interrupted format separately from a ready filesystem, retries the eligible unfinished format and only marks the data partition ready after successful formatting/mounting. New data partitions are labeled exFAT.

This addresses the provisioning failure path. Fresh-card first boot of this exact release remains a physical qualification item; an unknown existing data partition is not blindly reformatted.

[Implementation / details](https://github.com/dormancygrace/NanoKVM-OS/blob/9f65f9554142db4a61ba866b4498363f4f21d711/kvmapp/system/init.d/S01fs).

If you're still experiencing this issue, you're welcome to try it and let me know whether it resolves the problem on your setup.

It's currently in beta. Development testing has been on PCIe/UXC; Cube testing is still pending. Installing from stock firmware requires flashing a full SD image, so please back up your card first.

[Latest release](https://github.com/dormancygrace/NanoKVM-OS/releases/latest) · [Installation guide](https://github.com/dormancygrace/NanoKVM-OS/blob/main/docs/INSTALL.md)

Please include your board revision and reproduction steps when reporting results. Thanks!
